Modeling Source Change Verdicts (#415)
eiffel-community/eiffel: eiffel-syntax-and-usage/clm-source-change-approved.md

<!---
   Copyright 2025 Ericsson AB.
   For a full list of individual contributors, please see the commit history.

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
--->

# Source Change Approval Using Confidence Levels

_Source change approval_ is the process of signaling when a source change has been reviewed and approved according to project requirements. This document describes how to represent source change approval states using [EiffelConfidenceLevelModifiedEvent][CLM] events, providing a standardized way to communicate approval status using the Eiffel protocol.

> Source change approval is a critical gate in software development workflows, determining when changes are ready for integration. By modeling approval as confidence levels, teams can create flexible approval workflows that integrate naturally with existing Eiffel-based CI/CD pipelines.

Reviewer: ```suggestion
Source change approvals can be used as a trigger to execute certain test pipelines. Those pipelines could either consume serious amounts of hardware/network resources and thus should not run unless the source change is deemed interesting enough, or they could run in sensitive environments and should therefore not be executed unless the reviewer(s) say it's safe to run. By modeling approval as confidence levels, teams can create flexible approval workflows that integrate naturally with existing Eiffel-based CI/CD pipelines.

```

## Confidence Level as Approval State

The [EiffelConfidenceLevelModifiedEvent][CLM] provides a natural way to represent approval states by treating approval as a confidence level in the source change. This approach offers several advantages:

- **Semantic alignment**: Approval represents confidence in the quality and readiness of a change
- **Graduated states**: Can represent progressive approval states (pending, partial, full)
- **Standard tooling**: Existing Eiffel tools understand confidence level events
- **Flexible values**: Supports different approval schemes and requirements

## Ideas for Approval State Representations

To exemplify how the CLM event could be used, we show some alternatives.

### Simple Approval States

The most straightforward approach uses discrete approval states:

```json
{
  "data": {

>     "name": "review-approval",

Reviewer: The CLM docs recommend camelCase on the name
```suggestion
    "name": "reviewApproval",

```

    "value": "APPROVED"
  }
}
```

> **Supported values:**

Reviewer: The schema does not limit what values are possible to use, so I don't think we should say 'Supported values'
```suggestion
**Proposed values:**

```

- `PENDING`: Review process initiated but not complete
- `APPROVED`: All required approvals received
- `REJECTED`: Change rejected during review

> - `INSUFFICIENT_REVIEWS`: Not enough reviews to meet requirements

Reviewer: In what scenario should PENDING be used and not INSUFFICIENT_REVIEWS or vice versa? To me, they are mostly the same. I propose to remove the last value.
